### PR TITLE
prefit for CQR

### PR DESCRIPTION
**Pull request by LacombeLouis**
# Description

Providing a ``cv == "prefit`` option for the CQR. This means adding some additional tests and trusting the user to correctly set the quantiles for their method. This is to fix issue #213.

## Type of change

Please remove options that are irrelevant.

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- Re-create the paper by using the algorithms used in his paper and use them prefit.

# Checklist

- [ ] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [ ] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [ ] Linting passes successfully : `make lint`
- [ ] Typing passes successfully : `make type-check`
- [ ] Unit tests pass successfully : `make tests`
- [ ] Coverage is 100% : `make coverage`
- [ ] Documentation builds successfully : `make doc`

### CODE REVIEW

File: mapie/quantile_regression.py [hunk 1]
```diff
@@ -138,7 +139,11 @@ class MapieQuantileRegressor(MapieRegressor):
 
     def __init__(
         self,
-        estimator: Optional[Union[RegressorMixin, Pipeline]] = None,
+        estimator: Optional[
+            Union[RegressorMixin, Pipeline, List[
+                    Union[RegressorMixin, Pipeline]
+                ]]
+            ] = None,
```
Comment (vtaquet) [inline mapie/quantile_regression.py:146]: For sake of clarity, I would rearrange it to :
```
        estimator: Optional[
            Union[
                RegressorMixin,
                Pipeline,
                List[Union[RegressorMixin, Pipeline]]
            ]
        ] = None
```

File: mapie/quantile_regression.py [hunk 2]
```diff
@@ -462,65 +529,80 @@ def fit(
...
 
-        if isinstance(checked_estimator, Pipeline):
-            estimator = checked_estimator[-1]
+            y_calib_preds = np.full(
+                shape=(3, self.n_calib_samples),
+                fill_value=np.nan
+            )
+            for i, est in enumerate(estimator):
+                self.estimators_.append(est)
+                y_calib_preds[i] = est.predict(X_calib)
```
Comment (vtaquet) [inline mapie/quantile_regression.py:548]: As discussed yesterday, I think it's better to use `X` and `y` with `cv="prefit"` in order to make the class consistent with `MapieRegressor` and to simplify the use of `.fit`.

File: mapie/quantile_regression.py [hunk 3]
```diff
@@ -318,7 +323,7 @@ def _check_cv(
         """
         if cv is None:
             return "split"
-        if cv == "split":
+        if (cv == "split") or (cv == "prefit"):
```
Comment (vtaquet) [inline mapie/quantile_regression.py:326]: Add "prefit" in documentation (line 306).

File: mapie/quantile_regression.py [hunk 4]
```diff
@@ -462,65 +529,80 @@ def fit(
...
-            name_estimator
-        ]["alpha_name"]
-        for i, alpha_ in enumerate(alpha):
-            cloned_estimator_ = clone(checked_estimator)
-            params = {alpha_name: alpha_}
+            # Checks
+            self._check_parameters()
+            checked_estimator = self._check_estimator(self.estimator)
+            alpha = self._check_alpha(self.alpha)
+            self.cv = self._check_cv(cast(str, self.cv))
```
Comment (vtaquet) [inline mapie/quantile_regression.py:554]: This check needs to be placed at the very beginning of the method before the if/else statements.

File: mapie/quantile_regression.py [hunk 5]
```diff
@@ -462,65 +529,80 @@ def fit(
...
-            y_train
-        )
-        y_train = cast(NDArray, y_train)
-
         # Initialization
         self.estimators_: List[RegressorMixin] = []
+        if self.cv == "prefit":
+            estimator = cast(List, self.estimator)
+            self._prefit_params(estimator, X, y, X_calib, y_calib)
+            X_calib, y_calib = indexable(X_calib, y_calib)
```
Comment (vtaquet) [inline mapie/quantile_regression.py:537]: `X` and `y` instead of `X_calib` and `y_calib`.

File: mapie/quantile_regression.py [hunk 6]
```diff
@@ -462,65 +529,80 @@ def fit(
...
 
-        # Work
-        y_calib_preds = np.full(
-            shape=(3, self.n_calib_samples),
-            fill_value=np.nan
-        )
+            self.n_calib_samples = _num_samples(y_calib)
```
Comment (vtaquet) [inline mapie/quantile_regression.py:539]: `y` instead of `y_calib`.

**Review by vtaquet — COMMENTED** [review]
Very nice implementation Louis ! I have some minor comments about it. Please feel free to :
- take them into account in `quantile_regression.py`
- add new unit tests that checks `cv="prefit"` (you can have a look at the tests for `regression.py`)
- add a tutorial (using the California dataset ?) explaining how to use CQR with MAPIE

File: mapie/quantile_regression.py (hunk 3 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:326]: done

File: mapie/quantile_regression.py [hunk 7]
```diff
@@ -138,7 +139,11 @@ class MapieQuantileRegressor(MapieRegressor):
 
     def __init__(
         self,
-        estimator: Optional[Union[RegressorMixin, Pipeline]] = None,
+        estimator: Optional[
```
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/quantile_regression.py:142]: done

File: mapie/quantile_regression.py [hunk 8]
```diff
@@ -318,7 +323,7 @@ def _check_cv(
         """
         if cv is None:
             return "split"
-        if cv == "split":
+        if (cv == "split") or (cv == "prefit"):
             return cv
         else:
             raise ValueError(
```
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/quantile_regression.py:329]: done

File: mapie/quantile_regression.py [hunk 9]
```diff
@@ -397,6 +402,68 @@ def _check_calib_set(
         sample_weight_train = cast(ArrayLike, sample_weight_train)
         return X_train, y_train, X_calib, y_calib, sample_weight_train
 
+    def _prefit_params(
```
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/quantile_regression.py:405]: done

File: mapie/quantile_regression.py [hunk 10]
```diff
@@ -397,6 +402,68 @@ def _check_calib_set(
...
+            If less or more than three models are defined.
+        Warning
+            If X and y are defined, then warning that they are not used.
+        ValueError
+            If the calibration set is not defined.
+        Warning
+            If the alpha is defined, warns the user that it must be set
+            accordingly with the prefit estimators.
+        """
+        if len(estimator) == 3:
```
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/quantile_regression.py:443]: done

File: mapie/quantile_regression.py [hunk 11]
```diff
@@ -397,6 +402,68 @@ def _check_calib_set(
...
+                + " trained."
+            )
+        if X_calib is None or y_calib is None:
+            raise ValueError(
+                "You need to set the calibration dataset."
+            )
+        if self.alpha is not None:
+            warnings.warn(
+                "WARNING: The alpha that is set needs to be the same"
+                + " as the alpha of your prefitted model."
```
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/quantile_regression.py:464]: done

File: mapie/quantile_regression.py [hunk 12]
```diff
@@ -397,6 +402,68 @@ def _check_calib_set(
...
+        if X is not None and y is not None:
+            warnings.warn(
+                "WARNING: X and y will not be used as model is already"
+                + " trained."
+            )
+        if X_calib is None or y_calib is None:
+            raise ValueError(
+                "You need to set the calibration dataset."
+            )
+        if self.alpha is not None:
```
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/quantile_regression.py:461]: Not sure of the usefulness of this. Most important is to make sure the user knows how he has setup the estimators when using the prefit method. (and would only allow us to check for the ones we can already pass using the "split" method)

File: mapie/quantile_regression.py (hunk 5 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:537]: Changing line 536, but in line 537, we need to output `X_calib, y_calib` as `y_calib` is used later to calculate conformity scores.

File: mapie/quantile_regression.py (hunk 6 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:539]: Same reasoning as previous comment, so keeping it as `y_calib`.

File: mapie/quantile_regression.py (hunk 2 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:548]: Done, to the extent that in the fit method, they use `X` and `y`, however, in the variable namings, we still use `y_calib` later to computer conformity scores.

File: mapie/quantile_regression.py (hunk 4 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:554]: done

File: mapie/regression.py [hunk 13]
```diff
@@ -322,7 +322,7 @@ def _check_estimator(
             return LinearRegression()
         if not (hasattr(estimator, "fit") and hasattr(estimator, "predict")):
             raise ValueError(
-                "Invalid estimator. "
+                "Invalid estimator."
```
Comment (vtaquet) [inline mapie/regression.py:325]: Can use the new `check_estimator_fit_predict` function ?

File: mapie/quantile_regression.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (vtaquet) [reply mapie/quantile_regression.py]: Agreed, let's keep the warnings message then.

File: environment.doc.yml [hunk 14]
```diff
@@ -11,3 +11,4 @@ dependencies:
     - sphinx-gallery=0.10.1
     - sphinx_rtd_theme=1.0.0
     - typing_extensions=4.0.1
+    - lightgbm=3.1.1
```
Comment (vtaquet) [inline environment.doc.yml:14]: Add it in alphabetical order.

File: mapie/utils.py [hunk 15]
```diff
@@ -563,3 +563,26 @@ def check_defined_variables_predict_cqr(
             "WARNING: Alpha should not be specified in the prediction method\n"
             + "with conformalized quantile regression."
         )
+
+
+def check_estimator_fit_predict(
```
Comment (vtaquet) [inline mapie/utils.py:568]: This functions needs to have a dedicated test in `test_utils.py`.

File: examples/regression/1-quickstart/plot_prefit.py [hunk 16]
```diff
@@ -3,21 +3,26 @@
 Example use of the prefit parameter with neural networks
 ========================================================
 
-:class:`mapie.regression.MapieRegressor` is used to calibrate
-uncertainties for large models for which the cost of cross-validation
-is too high. Typically, neural networks rely on a single validation set.
+:class:`mapie.regression.MapieRegressor` and
+:class:`mapie.quantile_regression.MapieQuantileRegressor``
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:7]: Remove a `.

File: examples/regression/1-quickstart/plot_prefit.py [hunk 17]
```diff
@@ -3,21 +3,26 @@
...
 then compute residuals on a validation set with the `cv="prefit"` parameter.
 Finally, we evaluate the model with prediction intervals on a testing set.
 """
-import scipy
+
+
 import numpy as np
-from sklearn.model_selection import train_test_split
-from sklearn.neural_network import MLPRegressor
+from lightgbm import LGBMRegressor
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:19]: The use of `LGBMRegressor` is not compatible with the title of this example ;-). Besides, e wanted to use a MLP to emphasize the fact that MAPIE can be used with deep learning models. Use LGBM as a second model in addition to MLPs and change the title ?

File: examples/regression/1-quickstart/plot_prefit.py [hunk 18]
```diff
@@ -55,6 +63,37 @@ def f(x: NDArray) -> NDArray:
 y_pred_low, y_pred_up = y_pis[:, 0, 0], y_pis[:, 1, 0]
 coverage = regression_coverage_score(y_test, y_pred_low, y_pred_up)
 
+# Train models for MapieQuantileRegressor
+list_estimators = []
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:67]: Train the three models in a loop ?

File: examples/regression/1-quickstart/plot_prefit.py [hunk 19]
```diff
@@ -79,11 +118,25 @@ def f(x: NDArray) -> NDArray:
...
+    y_pred_low_cqr[order],
+    y_pred_up_cqr[order],
+    alpha=0.4,
+    label="prediction intervals CQR"
+)
 plt.title(
-    f"Target and effective coverages for "
-    f"alpha={alpha}: ({1 - alpha:.3f}, {coverage:.3f})"
+    f"Target and effective coverages for:\n "
+    f"QR alpha={alpha}: ({1 - alpha:.3f}, {coverage:.3f})\n"
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:138]: QR stands for quantile regression ? If so, I'm not sure the term is correct as you just used a LGBM fitted with a QR and then applied a split-conformal method with MAPIE.

File: mapie/tests/test_quantile_regression.py [hunk 20]
```diff
@@ -519,6 +519,120 @@ def test_linear_regression_results(strategy: str) -> None:
...
+def test_quantile_prefit_non_list() -> None:
+    """
+    Test that there is a list of estimators provided when cv='prefit'
+    is called for MapieQuantileRegressor.
+    """
+    with pytest.raises(
+        ValueError,
+        match=r".*Estimator for prefit must be an iterable object.*",
+    ):
+        not_an_iterable = 10
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:531]: You can parametrize the argument "not_an_iterable" to test different possibilities.

File: mapie/tests/test_quantile_regression.py [hunk 21]
```diff
@@ -519,6 +519,120 @@ def test_linear_regression_results(strategy: str) -> None:
...
+        )
+        mapie_reg.fit(
+            X_calib_toy,
+            y_calib_toy
+        )
+
+
+def test_quantile_prefit_three_estimators() -> None:
+    """
+    Test that there is a list of estimators three estimators provided for
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:544]: a list with three estimators ?

File: mapie/tests/test_quantile_regression.py [hunk 22]
```diff
@@ -519,6 +519,120 @@ def test_linear_regression_results(strategy: str) -> None:
...
+        )
+        mapie_reg.fit(
+            X_calib,
+            y_calib
+        )
+
+
+def test_prefit_no_fit_predict() -> None:
+    """
+    Check that the user is warned that the alphas need to be correctly set.
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:567]: Docstring does not correspond to the test.

File: mapie/regression.py (hunk 13 above)
Comment (LacombeLouis) [reply mapie/regression.py:325]: done

File: environment.doc.yml (hunk 14 above)
Comment (LacombeLouis) [reply environment.doc.yml:14]: done

File: mapie/tests/test_quantile_regression.py (hunk 20 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:531]: Do you have ideas of what other types of non-iterables I could test?

File: examples/regression/1-quickstart/plot_prefit.py (hunk 18 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:67]: done

File: mapie/tests/test_quantile_regression.py (hunk 21 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:544]: done

File: mapie/tests/test_quantile_regression.py (hunk 22 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:567]: done

File: mapie/tests/test_quantile_regression.py [hunk 23]
```diff
@@ -519,6 +519,120 @@ def test_linear_regression_results(strategy: str) -> None:
...
+        )
+        mapie_reg.fit(
+            X_calib,
+            y_calib
+        )
+
+
+def test_non_trained_estimator() -> None:
+    """
+    Check that the user is warned that the alphas need to be correctly set.
```
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:591]: done

File: mapie/tests/test_quantile_regression.py [hunk 24]
```diff
@@ -519,6 +519,120 @@ def test_linear_regression_results(strategy: str) -> None:
...
+            estimator=list_estimators,
+            cv="prefit",
+            alpha=0.3
+        )
+        mapie_reg.fit(
+            X_calib,
+            y_calib
+        )
+
+
```
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:635]: done

File: examples/regression/1-quickstart/plot_prefit.py (hunk 19 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:138]: changed

File: examples/regression/1-quickstart/plot_prefit.py (hunk 17 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:19]: changed

File: mapie/tests/test_quantile_regression.py [hunk 25]
```diff
@@ -519,6 +519,151 @@ def test_linear_regression_results(strategy: str) -> None:
...
+def test_prefit_and_non_prefit_equal() -> None:
+    """
+    Check that the user is warned that the alphas need to be correctly set.
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:638]: Docstring to update.

File: mapie/tests/test_quantile_regression.py [hunk 26]
```diff
@@ -519,6 +519,151 @@ def test_linear_regression_results(strategy: str) -> None:
...
+    """
+    list_estimators = []
+    alphas_ = [0.15, 0.85, 0.5]
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:641]: Possible to parametrize `alpha` (and test 2-3 values) and deduce the `alphas_` accordingly ?

File: mapie/tests/test_quantile_regression.py [hunk 27]
```diff
@@ -519,6 +519,151 @@ def test_linear_regression_results(strategy: str) -> None:
...
+
+    mapie_reg = MapieQuantileRegressor(
+        estimator=qt,
+        alpha=0.3
+    )
+    mapie_reg.fit(X_train, y_train, X_calib=X_calib, y_calib=y_calib)
+    y_pred, y_pis = mapie_reg.predict(X)
+
+    np.testing.assert_allclose(y_pred_prefit, y_pred)
+    np.testing.assert_allclose(y_pis_prefit, y_pis)
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:664]: Add a last test check that MAPIE returns the same results if the 3 estimators are given as a list, tuple, or a np.array ?

File: examples/regression/1-quickstart/plot_prefit.py [hunk 28]
```diff
@@ -0,0 +1,201 @@
...
+:class:`mapie.regression.MapieRegressor` and
+:class:`mapie.quantile_regression.MapieQuantileRegressor`
+are used to calibrate uncertainties for large models for
+which the cost of cross-validation is too high. Typically,
+neural networks rely on a single validation set.
+
+In this example, we first fit a neural network on the training set. We
+then compute residuals on a validation set with the `cv="prefit"` parameter.
+Finally, we evaluate the model with prediction intervals on a testing set.
+We will also show how to use the prefit method in the comformalized quantile
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:15]: typo "comformalized"

File: examples/regression/1-quickstart/plot_prefit.py [hunk 29]
```diff
@@ -0,0 +1,201 @@
...
+from matplotlib import pyplot as plt
+import scipy
+from sklearn.model_selection import train_test_split
+from sklearn.neural_network import MLPRegressor
+
+from mapie.regression import MapieRegressor
+from mapie.quantile_regression import MapieQuantileRegressor
+from mapie.metrics import regression_coverage_score
+from mapie._typing import NDArray
+import warnings
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:31]: Move this import before the mapie importations ?

File: examples/regression/1-quickstart/plot_prefit.py [hunk 30]
```diff
@@ -0,0 +1,201 @@
...
+
+alpha = 0.1
+
+##############################################################################
+# 1. Generate dataset
+# -----------------------------------------------------------------------------
+#
+# We start by defining a function that we will use to generate data. We then
+# add random noise y values. Then we split the dataset to have a training,
+# calibration and test set.
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:42]: to y values ?

File: examples/regression/1-quickstart/plot_prefit.py [hunk 31]
```diff
@@ -0,0 +1,201 @@
...
+    X_train_cal, y_train_cal, test_size=1 / 9
+)
+
+
+##############################################################################
+# 2. Pre-train models
+# -----------------------------------------------------------------------------
+#
+# For this example, we will train a MLPRegressor for
+# :class:`mapie.regression.MapieRegressor` and multiple LGBMRegressor in the
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:70]: "in the" to remove ?

File: examples/regression/1-quickstart/plot_prefit.py [hunk 32]
```diff
@@ -0,0 +1,201 @@
...
+# with a quantile objective as this is a requirement to perform conformalized
+# quantile regression using
+# :class:`mapie.quanitle_regression.MapieQuantileRegressor`. Note that the
+# three estimators need to be trained at quantile values of
+# $(\alpha/2, 1-(\alpha/2), 0.5)$.
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:75]: :math:`` instead of $$

File: examples/regression/1-quickstart/plot_prefit.py [hunk 33]
```diff
@@ -0,0 +1,201 @@
...
+
+
+# Train a MLPRegressor for MapieRegressor
+est_mlp = MLPRegressor(activation="relu", random_state=1)
+est_mlp.fit(X_train.reshape(-1, 1), y_train)
+
+# Train LGBMRegressor models for MapieQuantileRegressor
+list_estimators_cqr = []
+for alpha_ in [alpha/2, (1-(alpha/2)), 0.5]:
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:84]: spaces before/after -, /

File: examples/regression/1-quickstart/plot_prefit.py [hunk 34]
```diff
@@ -0,0 +1,201 @@
...
+    )
+    estimator_.fit(X_train.reshape(-1, 1), y_train)
+    list_estimators_cqr.append(estimator_)
+
+
+##############################################################################
+# 3. Using MAPIE to calibrate the models
+# -----------------------------------------------------------------------------
+#
+# We will now proceed to calibrate the models using MAPIE. This means using
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:97]: To this aim, we set `cv="prefit"`

File: examples/regression/1-quickstart/plot_prefit.py [hunk 35]
```diff
@@ -0,0 +1,201 @@
...
+    y_pis_cqr[:, 0, 0],
+    y_pis_cqr[:, 1, 0]
+)
+
+
+##############################################################################
+# 4. Plots
+# -----------------------------------------------------------------------------
+#
+# In order to view the results shown above, we will plot each othe predictions
```
Comment (vtaquet) [inline examples/regression/1-quickstart/plot_prefit.py:127]: typo other ?

File: mapie/tests/test_quantile_regression.py (hunk 20 above)
Comment (vtaquet) [reply mapie/tests/test_quantile_regression.py]: A float, a string ?

File: mapie/tests/test_quantile_regression.py (hunk 25 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:638]: done

File: mapie/tests/test_quantile_regression.py (hunk 26 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:641]: done

File: mapie/tests/test_quantile_regression.py (hunk 27 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:664]: done

File: examples/regression/1-quickstart/plot_prefit.py (hunk 28 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:15]: done

File: examples/regression/1-quickstart/plot_prefit.py (hunk 29 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:31]: done

File: examples/regression/1-quickstart/plot_prefit.py (hunk 30 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:42]: to the dependent variable?

File: examples/regression/1-quickstart/plot_prefit.py (hunk 31 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:70]: done

File: examples/regression/1-quickstart/plot_prefit.py (hunk 32 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:75]: done

File: examples/regression/1-quickstart/plot_prefit.py (hunk 33 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:84]: done

File: examples/regression/1-quickstart/plot_prefit.py (hunk 34 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:97]: done

File: examples/regression/1-quickstart/plot_prefit.py (hunk 35 above)
Comment (LacombeLouis) [reply examples/regression/1-quickstart/plot_prefit.py:127]: done

File: mapie/tests/test_quantile_regression.py (hunk 20 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py]: A string is iterable :(. I have added float. Also do we need to add this test both in `test_quantile_regression.py` and `test_utils.py`?

File: mapie/utils.py (hunk 15 above)
Comment (LacombeLouis) [reply mapie/utils.py:568]: Done, does this need to be added to both `test_quantile_regression.py` and `test_utils.py`?

File: mapie/tests/test_quantile_regression.py [hunk 36]
```diff
@@ -519,6 +519,199 @@ def test_linear_regression_results(strategy: str) -> None:
...
+
+
+@pytest.mark.parametrize("alpha", [0.05, 0.1, 0.2, 0.3])
+def test_prefit_different_type_list_tuple_array(alpha: float) -> None:
+    """
+    Check that as long as the estimators are given in a Iterable object,
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:672]: Simplify to "Check that the type of Iterable (list, np.array, tuple) to estimators gives similar results." ?

File: mapie/tests/test_quantile_regression.py (hunk 20 above)
Comment (vtaquet) [reply mapie/tests/test_quantile_regression.py]: Only one in `test_quantile_regression` is enough.

File: examples/regression/1-quickstart/plot_prefit.py (hunk 30 above)
Comment (vtaquet) [reply examples/regression/1-quickstart/plot_prefit.py]: You forgot "to" in the sentence.

File: mapie/tests/test_quantile_regression.py (hunk 36 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:672]: done

File: mapie/utils.py (hunk 15 above)
Comment (LacombeLouis) [reply mapie/utils.py:568]: done

**Review by vtaquet — APPROVED** [review]
Looks good to me ! Thanks Louis !

**Pull request merged**